### ISSUE TITLE
[Reverse C Interop] Memory leaks detected when Kotlin called on background thread

### ISSUE THREAD

**Issue by alusch** [JetBrains/kotlin-native#3705]
knlibrary.kt:
```
package knlibrary

class Thing {}
```

test.cpp:
```
#include <thread>
#include "knlibrary_api.h"

int main()
{
    auto t = std::thread([]
    {
        auto lib = knlibrary_symbols();
        auto knlibrary = lib->kotlin.root.knlibrary;

        auto thing = knlibrary.Thing.Thing();
        lib->DisposeStablePointer(thing.pinned);
    });
    t.join();
    
    return 0;
}
```

run.cmd:
```
@echo off

call kotlinc-native -g knlibrary.kt -produce dynamic -o knlibrary.dll

lib.exe /def:knlibrary.def /out:knlibrary.lib /machine:x64
cl.exe test.cpp knlibrary.lib
test.exe
```

This assumes you have the 64-bit cl.exe on your path (e.g. using the VS x64 Native Tools Command Prompt), as well as the Kotlin and Java bin directories.

I expect no output since I'm disposing the one pointer I get from Kotlin. However, I actually see
```
Memory leaks detected, 8 objects leaked!
Use `Platform.isMemoryLeakCheckerActive = false` to avoid this check.
```

when testing on Kotlin 1.3.61. Building a compiler with `#define TRACE_MEMORY 1` shows the following details:
```
*** Memory leaks, leaked 8 containers ***
BLACK regular frozen container 000001d148233c30 with 1 objects rc=1
    000001d148233c38 kotlin.Array
GREEN regular frozen container 000001d148233b10 with 1 objects rc=1
    000001d148233b18 kotlin.reflect.KVariance
GREEN regular frozen container 000001d148233bd0 with 1 objects rc=1
    000001d148233bd8 kotlin.reflect.KVariance
GREEN regular frozen container 000001d148233c00 with 1 objects rc=1
    000001d148233c08 kotlin.reflect.KVariance
GREEN regular frozen container 000001d148233bd0 with 1 objects rc=1
    000001d148233bd8 kotlin.reflect.KVariance
GREEN regular frozen container 000001d148233b10 with 1 objects rc=1
    000001d148233b18 kotlin.reflect.KVariance
GREEN regular frozen container 000001d1482333e0 with 1 objects rc=1
    000001d1482333e8 kotlin.collections.EmptyList
BLACK regular frozen container 000001d148233320 with 1 objects rc=1
    000001d148233328 kotlin.reflect.KVariance.$OBJECT
BLACK regular frozen container 000001d1482336a0 with 1 objects rc=1
    000001d1482336a8 kotlin.reflect.KTypeProjection.Companion
BLACK regular frozen container 000001d148233ae0 with 1 objects rc=1
    000001d148233ae8 kotlin.reflect.KTypeProjection
BLACK regular frozen container 000001d148233ae0 with 1 objects rc=1
    000001d148233ae8 kotlin.reflect.KTypeProjection
GREEN regular frozen container 000001d148233c00 with 1 objects rc=1
    000001d148233c08 kotlin.reflect.KVariance
```

Removing the thread and doing the work within `main` does not report any leaks, including if I deliberately leak `thing`, so I'm not sure the leak detector is actually running in that case. It's also unclear to me if these are legitimate leaks or false positives. Thanks for taking a look!

**Comment by olonho** [comment]
This shall be fixed with 1.3.70, please recheck with 1.3.70-eap-19

**Comment by artdfel** [comment]
This issue seems to be fixed in 1.3.70, as expected. Also, I was unable to reproduce with 1.4.10. Therefore I'm closing this one and asking to report any other problems to https://kotl.in/issue, see #4079 for the details.
